pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,186 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"buf.build/gen/go/open-feature/flagd/grpc/go/sync/v1/syncv1grpc"
+	v1 "buf.build/gen/go/open-feature/flagd/protocolbuffers/go/sync/v1"
+
+	"github.com/open-feature/flagd/pkg/logger"
+	"github.com/open-feature/flagd/pkg/sync"
+	"google.golang.org/grpc"
+)
+
+const (
+	// Prefix for GRPC URL inputs. GRPC does not define a prefix through standard. This prefix helps to differentiate
+	// remote URLs for REST APIs (i.e - HTTP) from GRPC endpoints.
+	Prefix = "grpc://"
+
+	// Connection retry constants
+	// Back off period is calculated with backOffBase ^ #retry-iteration. However, when #retry-iteration count reach
+	// backOffLimit, retry delay fallback to constantBackOffDelay
+	backOffLimit         = 3
+	backOffBase          = 4
+	constantBackOffDelay = 60
+)
+
+type Sync struct {
+	Target     string
+	ProviderID string
+	Logger     *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	options := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// initial dial and connection. Failure here must result in a startup failure
+	dial, err := grpc.DialContext(ctx, g.Target, options...)
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing grpc connection: %s", err.Error()))
+		return err
+	}
+
+	serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+	syncClient, err := serviceClient.SyncFlags(ctx, &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error calling streaming operation: %s", err.Error()))
+		return err
+	}
+
+	// initial stream listening
+	err = g.handleFlagSync(syncClient, dataSync)
+	if err != nil {
```

Comment:
err can never be nil, could either leave a comment indicating this or remove the nil check